text-spacing: text-autospace: TextSpacing context refactoring

vitorroriz opened the pull request:
#### b0234689a4e50dd76e415723ac7b120b545691a6
<pre>
TextSpacing context refactoring
<a href="https://bugs.webkit.org/show_bug.cgi?id=283005">https://bugs.webkit.org/show_bug.cgi?id=283005</a>
<a href="https://rdar.apple.com/139743029">rdar://139743029</a>

Reviewed by Alan Baradlay.

This should produce no behavioral changes.

Currently we transport TextSpacing information about spacing
between element boundaries within IFC as `InlineBoxBoundaryTextSpacings`.

In this patch, we are wrapping up this data into a more generic data
structure called TextSpacingContext. This is because we need to transport
further data for addressing the trimming of spacing during line-breaking properly.
This extra data transported is called here `TrimmableTextSpacings`, which is always
default constructed and unused in this patch, until it gets implemented by another
patch in the future.

This approach also allow us to expand the data being transmitted
for text-spacing purposes if needed later.

* Source/WebCore/layout/formattingContexts/inline/InlineContentCache.h:
(WebCore::Layout::InlineContentCache::inlineBoxBoundaryTextSpacings const):
(WebCore::Layout::InlineContentCache::setInlineBoxBoundaryTextSpacings):
(WebCore::Layout::InlineContentCache::trimmableTextSpacings const):
(WebCore::Layout::InlineContentCache::setTrimmableTextSpacings):
(WebCore::Layout::InlineContentCache::textSpacingContext const):
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
(WebCore::Layout::InlineFormattingContext::layout):
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
(WebCore::Layout::LineBuilder::LineBuilder):
(WebCore::Layout::LineBuilder::candidateContentForLine):
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.h:
* Source/WebCore/layout/formattingContexts/inline/InlineLineTypes.h:

Canonical link: <a href="https://commits.webkit.org/286548@main">https://commits.webkit.org/286548@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ffd534b92bb305269dc7972b929b28da29d3102

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76170 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55200 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29072 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80673 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27439 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78286 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64344 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3496 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59721 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17872 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79237 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49624 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65416 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40070 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47020 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22899 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25764 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68153 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23234 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82134 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3542 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2344 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67950 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3696 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65384 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67260 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16798 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11219 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9332 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3490 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6297 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3513 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/6940 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5271 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8ffd534b92bb305269dc7972b929b28da29d3102. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/36549

brentfulgham reviewed (commented):
I think these changes look fine mechanically, but I'd prefer to get @alanbaradlay or @sammygill 's opinion on whether this change matches their ideas for the overall architecture of the `InlineFormattingContext`.

If neither is available to review, I can sign off, but I'd like to defer to them for now.

For example: Should these trim/spacing concepts just be part of the existing `InlineBoxBoundaryTextSpacings` structure?

sammygill commented on Source/WebCore/layout/formattingContexts/inline/InlineLineTypes.h:
```diff
@@ -33,6 +33,12 @@ namespace Layout {
+using TrimmableTextSpacings = InlineBoxBoundaryTextSpacings;
+
+struct TextSpacingContext {
+    InlineBoxBoundaryTextSpacings inlineBoxBoundaryTextSpacings;
+    TrimmableTextSpacings trimmableTextSpacings;
```
I think the main question/concern I have is in regards to the relationship between these two fields within a `TextSpacingContext`.  

Are trimmable text spacings supposed to be a subset of the inline box boundary text spacings? If so, I would imagine it would be a bit tricky to make sure that these two are always in a valid state at any given point in time as it seems like these two things can get mutated arbitrarily.  Should we maybe consider changing what an “InlineBoxBoundaryTextSpacing" is and have something on it that indicates whether it is trimmable or not?

vitorroriz replied:
These are used independently. One is for identifying spacing between box boundaries, the other is to identify trimmable spacing between text runs.

`InlineBoxBoundaryTextSpacings` maps inline item indexes to spacing. It serves to identify start boxes that mark spacing between boxes boundaries. If we have start box with an entry here, this means that we have to add spacing between this start box and the previous box. We need to account for it during layout and to reproduce such spacing among 2 boxes during display. Therefore, this serve for identifying spacing between box boundaries such that we can both: add these spacing into layout/display and trim this spacing during line breaking if needed.

`TrimmableTextSpacings` map inline item indexes to spacing. Text content can be split into multiple different runs text items. This structure serves to identify spacing that was added to a text item because of its adjacency to another text item. For example, imagine we have adjacent text items A and B, and the last character of A and the first character of B require spacing to be added between them. This spacing is always added as a leading spacing to B. However, during line breaking, if B gets placed on a different line as A, they are no longer adjacent and we must trim such spacing. We don't use this map for actually adding the spacing, just for trimming it. Therefore this maps serves only to identify trimmable spacing between text runs. Maybe this can be called `TrimmableTextSpacingBetweenTextItems` or something.

vitorroriz committed:
TextSpacing context refactoring
https://bugs.webkit.org/show_bug.cgi?id=283005
rdar://139743029

Reviewed by Alan Baradlay.

This should produce no behavioral changes.

Currently we transport TextSpacing information about spacing
between element boundaries within IFC as `InlineBoxBoundaryTextSpacings`.

In this patch, we are wrapping up this data into a more generic data
structure called TextSpacingContext. This is because we need to transport
further data for addressing the trimming of spacing during line-breaking properly.
This extra data transported is called here `TrimmableTextSpacings`, which is always
default constructed and unused in this patch, until it gets implemented by another
patch in the future.

This approach also allow us to expand the data being transmitted
for text-spacing purposes if needed later.

* Source/WebCore/layout/formattingContexts/inline/InlineContentCache.h:
(WebCore::Layout::InlineContentCache::inlineBoxBoundaryTextSpacings const):
(WebCore::Layout::InlineContentCache::setInlineBoxBoundaryTextSpacings):
(WebCore::Layout::InlineContentCache::trimmableTextSpacings const):
(WebCore::Layout::InlineContentCache::setTrimmableTextSpacings):
(WebCore::Layout::InlineContentCache::textSpacingContext const):
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
(WebCore::Layout::InlineFormattingContext::layout):
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
(WebCore::Layout::LineBuilder::LineBuilder):
(WebCore::Layout::LineBuilder::candidateContentForLine):
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.h:
* Source/WebCore/layout/formattingContexts/inline/InlineLineTypes.h:

Canonical link: https://commits.webkit.org/286548@main

webkit-commit-queue commented:
Committed 286548@main (b0234689a4e5): <https://commits.webkit.org/286548@main>

Reviewed commits have been landed. Closing PR #36549 and removing active labels.
